### ISSUE TITLE
ServiceProviderExtensions missing GetServices and GetServices<T> extension method

### ISSUE THREAD

**Issue by Tasteful** [aspnet/DependencyInjection#241]
The `ServiceProviderExtensions` contains methods for `GetRequiredServices`/`GetRequiredServices<T>`. Equivalent method for `GetServices`/`GetServices<T>` without the `providers.Any()` check that always returning an empty `IEnumerable<T>` when non existings instances is registered in the container.

**Comment by henkmollema** [comment]
+1

**Comment by Eilon** [comment]
The final set of methods we want to have are:

```c#
// Keep these:
GetService() // comes from IServiceProvider interface
GetService<T>()

GetRequiredService()
GetRequiredService<T>()

// Add these:
GetServices()
GetServices<T>()

// Delete these:
GetRequiredServices()
GetRequiredServices<T>()
```

**Comment by henkmollema** [comment]
@Eilon I could look at this if you'd like.

**Comment by henkmollema** [comment]
Why drop the `GetRequiredServices` btw?

**Comment by rynowak** [comment]
> Why drop the GetRequiredServices btw?

We couldn't think of a use-case where you need 1 or more of something. Given the number of methods that already exist we wanted to make it easier to pick the one you want.

**Comment by henkmollema** [comment]
When applying the mediator pattern, I actually register multiple implementations of `INotificationHandler<TNotification>` and retrieve them using `GetRequiredServices()`.I'll be fine with `GetServices`, however.

Is that going to be implemented using `GetRequiredService()` or just with `GetService()`?

**Comment by pranavkm** [comment]
`GetServices` would do what `GetRequiredServices` does without the `Any` check:

```c#
public static IEnumerable<T> GetRequiredServices<T>([NotNull] this IServiceProvider provider)
 => provider.GetRequiredService<IEnumerable<T>>();
```

The plan is to do `Any` check on the consumer if required.

**Comment by pranavkm** [comment]
Fixed in e490facfd40d46ed73b8c601c9c731f4e251dead
